extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java

```diff
@@ -185,47 +195,43 @@ void validateConfigInjectionPoints(ValidationPhaseBuildItem validationPhase,
+                // @Observes @Initialized(ApplicationScoped.class) requires validation at static init
+                if (injectionPoint.getTarget().kind().equals(METHOD)
+                        && observerMethods.contains(injectionPoint.getTarget().asMethod())) {
+                    configProperties
+                            .produce(ConfigPropertyBuildItem.staticInit(propertyName, injectedType, propertyDefaultValue));
+                }
+
+                configProperties.produce(ConfigPropertyBuildItem.runtimeInit(propertyName, injectedType, propertyDefaultValue));
             }
         }
     }
 
     @BuildStep
-    @Record(RUNTIME_INIT)
-    void validateConfigValues(ConfigRecorder recorder, List<ConfigPropertyBuildItem> configProperties,
-            BeanContainerBuildItem beanContainer, BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
-        // IMPL NOTE: we do depend on BeanContainerBuildItem to make sure that the BeanDeploymentValidator finished its processing
-
-        // the non-primitive types need to be registered for reflection since Class.forName is used at runtime to load the class
-        for (ConfigPropertyBuildItem item : configProperties) {
-            Type requiredType = item.getPropertyType();
-            String propertyType = requiredType.name().toString();
-            if (requiredType.kind() != Kind.PRIMITIVE) {
-                reflectiveClass.produce(ReflectiveClassBuildItem.builder(propertyType).build());
-            }
-        }
+    @Record(STATIC_INIT)
+    void validateStaticInitConfigProperty(
+            ConfigRecorder recorder,
+            List<ConfigPropertyBuildItem> configProperties,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
 
-        Set<ConfigValidationMetadata> propertiesToValidate = new HashSet<>();
-        for (ConfigPropertyBuildItem configProperty : configProperties) {
-            String rawTypeName = configProperty.getPropertyType().name().toString();
-            List<String> actualTypeArgumentNames = Collections.emptyList();
-            if (configProperty.getPropertyType().kind() == Kind.PARAMETERIZED_TYPE) {
-                List<Type> argumentTypes = configProperty.getPropertyType().asParameterizedType().arguments();
-                actualTypeArgumentNames = new ArrayList<>(argumentTypes.size());
-                for (Type argumentType : argumentTypes) {
-                    actualTypeArgumentNames.add(argumentType.name().toString());
-                    if (argumentType.kind() != Kind.PRIMITIVE) {
-                        reflectiveClass.produce(ReflectiveClassBuildItem.builder(argumentType.name().toString())
-                                .build());
-                    }
-                }
+        recorder.validateConfigProperties(
+                configProperties.stream()
+                        .filter(p -> p.getExecutionMode().equals(ExecutionMode.STATIC_INIT))
```

Comment:
Just nitpicking, but if we add `ConfigPropertyBuildItem#isStaticInit()` then `configProperties.stream().filter(ConfigPropertyBuildItem::isStaticInit)` could be more readable...